typesetters/base.lua

```diff
@@ -912,6 +913,177 @@ function typesetter.leadingFor (_, vbox, previous)
   end
 end
 
+-- Beggining of liner logic (contructs spanning over several lines)
+
+-- These two special nodes are used to track the current liner entry and exit.
+-- As Sith Lords, they are always two: they are local here, so no one can
+-- use one alone and break the balance of the Force.
+local linerEnterNode = pl.class(SILE.nodefactory.hbox)
+function linerEnterNode:_init(name, outputMethod)
+  SILE.nodefactory.hbox._init(self)
+  self.outputMethod = outputMethod
+  self.name = name
+  self.is_enter = true
+end
+function linerEnterNode:clone()
+  return linerEnterNode(self.name, self.outputMethod)
+end
+function linerEnterNode:outputYourself ()
+  SU.error("A liner enter node " .. tostring(self) .. "'' made it to output", true)
+end
+function linerEnterNode:__tostring ()
+  return "+L[" .. self.name .. "]"
+end
+local linerLeaveNode = pl.class(SILE.nodefactory.hbox)
+function linerLeaveNode:_init(name)
+  SILE.nodefactory.hbox._init(self)
+  self.name = name
+  self.is_leave = true
+end
+function linerLeaveNode:clone()
+  return linerLeaveNode(self.name)
+end
+function linerLeaveNode:outputYourself ()
+  SU.error("A liner leave node " .. tostring(self) .. "'' made it to output", true)
+end
+function linerLeaveNode:__tostring ()
+  return "-L[" .. self.name .. "]"
+end
+
+local linerBox = pl.class(SILE.nodefactory.hbox)
+function linerBox:_init (name, outputMethod)
+  SILE.nodefactory.hbox._init(self)
+  self.width = SILE.length()
+  self.height = SILE.length()
+  self.depth = SILE.length()
+  self.name = name
+  self.inner = {}
+  self.outputYourself = outputMethod
+end
+function linerBox:append (node)
+  self.inner[#self.inner+1] = node
+  self.width = self.width + node.width
```

Comment:
Now breaks here with the manual at `documentation/c08-language.sil:213:3`

The manual is our real test case ^^

Reply:
> The manual is our real test case ^^

Well and autodoc with colors enabled, obviously.

Of course, it ought to work with `node.width:absolute()` here...

**But** I really thought everything was _already_ "absolutized" at this point (for Liang line breaking etc. to work with points everywhere), so wanted to understand the case.

Minimal scenario extracted from that error (in the SILE manual, located deep in an autodoc construct in the ruby package, but digging at the core of it)
```
\begin{document}
\use[module=packages.color]
\language[main=ja]
\color[color=red]{xxx}
\end{document}
```

Offending non-absolutized node: `K<0pt plus 0.25zw>`

Very interesting and enlightening. Japanese is one of our few cases with a fully custom node maker (not using the ICU-based one) and specific rules inserting such kerns in zw.

Reply:
When fixing JA don't think *too hard* about making it right, just make it work like it currently does. Which is poorly. There is a huge project needed to fix the CJK stack with an overhaul (and even a big $$$ out there to get it done!). we know the sequence of things being done is topsy-turvey so I don't feel too bad about just getting it monkey patched to get the job done for now knowing that any *real* fixes are going to be ... deep cutting.